### ISSUE TITLE
Docs should mention how the `--records-per-batch` option helps when processing streaming records

### ISSUE THREAD

**Issue by zmajeed** [johnkerl/miller#1178]
Thanks for the all the hard work and great features you've put into mlr!

I've run into an issue trying to use mlr for log tailing with `tail -f`, `kubectl logs -f --tail 0` and similar commands that continuously tail and stream lines

As of mlr 6.6 - tested with latest source on main - mlr processes records in batches of `const DEFAULT_RECORDS_PER_BATCH = 500` from `internal/pkg/cli/option_types.go`

```bash
$ mlr --version
mlr 6.6.0-dev
```

This creates a delay in seeing records appear when mlr is used in a pipeline with `tail -f`

Here's a simple test with JSON records - that's the format I'm interested in using mlr for

First simulate a stream of JSON records
```bash
# bash code generates batchSize JSON records per second
batchSize=1
for((;;)); do
  printf '{"hello": "world"}\n%.0s' $(seq $batchSize)
  sleep 1
done > records.json&
```

Then tail the stream and process records with mlr
```bash
time tail -f records.json |
mlr --json --no-jlistwrap --fflush cat |
IFS= read -r
```

Time for first record to appear:

| batchSize | time for first record to appear (seconds) |
| ---------- | --------------------------------- |
| 1              |  495.5 |
| 10            | 48.6 |
| 100          | 4.8   |
| 500         | 0.6  |

Note the `--fflush` option does not affect the record processing batch size - I use it to prevent additional delays because mlr is in the middle of the test pipeline and does not directly write to the terminal

What do you think of one or more commandline options for batching and writing a number of records with clearly stated defaults?

My test suggests an unpleasant workaround for now - insert a number of fake records into the log tail stream for each real record to meet mlr's record batch size - then remove the fake records before final output

```bash
time tail -f records.json |
# inject fake empty records for mlr to process and discard
while IFS= read -r line; do
  echo "$line"
  printf '{}\n%.0s' {1..500}
done |
mlr --json --no-jlistwrap --fflush skip-trivial-records then cat |
IFS= read -r
```

Now the first record for `batchSize = 1` appears in 0.5 seconds

Also note trying `stdbuf` to override buffering does not apply to this issue - and `stdbuf` is not available on Windows

**Comment by zmajeed** [comment]
Whoops - I just found the `--records-per-batch` option that takes care of this issue - I'll leave the issue open for anyone to comment on but feel free to close

**Comment by johnkerl** [comment]
Hi @zmajeed thank you for the _great_ write-up and sorry for my very long delay in replying. :( 

Indeed as you found this is due to `--records-per-batch`.

And also you also found this is not the usual line-buffering thing (`--fflush`, `stdbuf`) so often encountered in text processing. It's something Miller-specific and Go-specific.

The background (if you're interested) is: one of the [many reasons](https://github.com/johnkerl/miller/blob/main/README-dev.md) I ported Miller from C to Go was Go's use of channels and goroutines. These seemed like a perfect fit for Miller's [record-streaming architecture](https://miller.readthedocs.io/en/latest/reference-main-then-chaining/) -- one goroutine to read records, one per verb, and one to write records. Easy-peasy. Problem, though, that I didn't find until (sadly) quite deep in -- I didn't start at-scale testing as early as I should have -- when I put one record on each channel write/read the Go runtime spent a shockingly large amount of time _just scheduling goroutines_ 😞. It was pretty bad performance-wise. So I did the batching with default 500 records per batch -- each goroutine does a channel write/read only on batches of 500 records (or less, if at end of data.)

Anyway the upshot is (a) For performance I have no way to _not_ do this in the default case. And honestly I have a little buyer's remorse about using Go, as the runtime penality for dynamic allocation and GC is also (as measured by profiling) non-trivial. Maybe if I had it to do all over again I would have chosen another language. Anyway here we are. And (b) I failed -- flat-out failed -- to realize (and document) the (very obvious in retrospect) impact on `tail -f` contexts.

So you are spot-on, this needs a good write-up, and I'm sorry I didn't see this lack sooner.
